File: tutorial/java/Server/src/main/java/systems/zlink/tutorial/server/actors/Player.java
package systems.zlink.tutorial.server.actors;

import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.actors.ZLinkActorContext;

// --8<-- [start:actor-class]
// A player is addressed by its own id and carries state that outlives any one
// connection. Like a room, its messages run one at a time.
public final class Player implements ZLinkActor {

    private final ZLinkActorContext context;

    private String nickname = "anonymous";

    public Player(ZLinkActorContext context) {
        this.context = context;
    }

    @Override
    public ZLinkActorContext context() {
        return context;
    }

    public String nickname() {
        return nickname;
    }

    public void rename(String value) {
        nickname = value;
    }
}
// --8<-- [end:actor-class]

File: tutorial/kotlin/Server/src/main/kotlin/systems/zlink/tutorial/server/actors/Player.kt
package systems.zlink.tutorial.server.actors

import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import systems.zlink.framework.actors.ZLinkActor
import systems.zlink.framework.actors.ZLinkActorContext
import systems.zlink.framework.actors.ZLinkActorFactory

// --8<-- [start:actor-class]
// A player is addressed by its own id and carries state that outlives any one
// connection. Like a room, its messages run one at a time.
class Player(private val actorContext: ZLinkActorContext) : ZLinkActor {

    var nickname: String = "anonymous"
        private set

    override fun context(): ZLinkActorContext = actorContext

    fun rename(value: String) {
        nickname = value
    }
}

// --8<-- [end:actor-class]

// --8<-- [start:actor-factory]
// The Framework creates players through this factory rather than by calling a
// constructor, so dependencies can be injected here.
class PlayerFactory : ZLinkActorFactory {
    override fun create(context: ZLinkActorContext): CompletionStage<ZLinkActor> =
        CompletableFuture.completedFuture(Player(context))
}
// --8<-- [end:actor-factory]
